src/datasets/builder.py

```diff
@@ -432,6 +439,25 @@ def __setstate__(self, d):
     def manual_download_instructions(self) -> Optional[str]:
         return None
 
+    def _has_legacy_cache(self) -> bool:
+        """Check for the old cache directory template {cache_dir}/{namespace}___{builder_name}"""
+        if (
+            self.__module__.startswith("datasets.")
+            and not is_remote_url(self._cache_dir_root)
+            and self.repo_id is not None
+        ):
+            namespace = self.repo_id.split("/")[0]
+            legacy_cache_dir = os.path.join(self._cache_dir_root, f"{namespace}___{self.name}")
```

Comment:
This makes `_has_legacy_cache()` return `True` for all the datasets with the `parquet` builder if `<namespace>___parquet` directory just exists, independently of the dataset name and config name, which leads to replacing `config_name` and `self.dataset_name` and then breaking all the new cache structure (saving new dataset to the old-like cache structure instead of new-like). 

By the way, I think we should not replace `config_name` and `self.dataset_name` with their legacy versions in any case and just use them to check for old cache dirs, to have the correct `builder.info` object in the end, wdyt?

Reply:
maybe the builder info can be populated before they are replaced ?